modules/shortcodes/instagram.php

```diff
@@ -161,8 +163,63 @@ function jetpack_instagram_handler( $matches, $atts, $url ) {
 		$url_args['hidecaption'] = 'true';
 	}
 
-	$url = esc_url_raw( add_query_arg( $url_args, 'https://api.instagram.com/oembed/' ) );
+	$use_cache     = jetpack_instagram_use_cache( $matches, $atts, $passed_url );
+	$cache_key     = 'oembed_response_body_' . md5( add_query_arg( $url_args, 'https://api.instagram.com/oembed/' ) );
+	$response_body = $use_cache
+		? wp_cache_get( $cache_key, 'instagram_embeds' )
+		: false;
 
+	if ( ! $response_body ) {
+		$response_body = jetpack_instagram_fetch_embed( $url_args );
+	}
+
+	if ( is_wp_error( $response_body ) || empty( $response_body->html ) ) {
+		jetpack_instagram_output_errored_embed( $url );
+		return;
+	}
+
+	if ( $use_cache ) {
+		wp_cache_set(
+			$cache_key,
+			$response_body,
+			'instagram_embeds',
+			HOUR_IN_SECONDS + wp_rand( 0, HOUR_IN_SECONDS )
+		);
+	}
+
+	wp_enqueue_script(
+		'jetpack-instagram-embed',
+		Assets::get_file_url_for_environment( '_inc/build/shortcodes/js/instagram.min.js', 'modules/shortcodes/js/instagram.js' ),
+		array( 'jquery' ),
+		JETPACK__VERSION,
+		true
+	);
+
+	return $response_body->html;
+}
+
+/**
+ * Given a URL, will output an HTML comment and the linked URL.
+ *
+ * @param string $url The URL that was attempted to embed.
+ */
+function jetpack_instagram_output_errored_embed( $url ) {
+	?>
```

Comment:
Maybe we could provide more info to folks who can connect / reconnect? Something like:

```suggestion
	if (
		current_user_can( 'jetpack_connect' )
		&& ! ( new Status() )->is_offline_mode()
	) {
		echo '<div class="notice error" style="border-left:5px solid #dc3232;padding:1em;background-color:#f8f9f9;">';
		echo wp_kses(
			sprintf(
				/* translators: URL is a link to the Jetpack dashboard. */
				__( 'Your site is not connected to WordPress.com at the moment. Please <a href="%1$s">connect</a> to start using Instagram embeds.', 'jetpack' ),
				Jetpack::admin_url( 'page=jetpack#/dashboard' )
			),
			array(
				'a' => array(
					'href' => array(),
				),
			)
		);
		echo '</div>';
	}
	?>
```

Reply:
As discussed in Slack, it may be best to just not have any embed when the site is not connected to WordPress.com. 
p1597855236215600-slack-CBG1CP4EN